statusNotifierWatcher.js

```diff
@@ -79,42 +79,59 @@ const StatusNotifierWatcher = new Lang.Class({
         return bus_name + obj_path;
     },
 
+    _registerItem: function(service, bus_name, obj_path) {
+        let id = this._getItemId(bus_name, obj_path);
+
+        if (this._items[id]) {
+            Util.Logger.warn("Item "+id+" is already registered");
+            return;
+        }
+
+        Util.Logger.debug("Registering StatusNotifierItem "+id);
+
+        let indicator = new AppIndicator.AppIndicator(bus_name, obj_path);
+        let visual = new IndicatorStatusIcon.IndicatorStatusIcon(indicator);
+        indicator.connect('destroy', visual.destroy.bind(visual));
+
+        this._items[id] = indicator;
+
+        this._dbusImpl.emit_signal('StatusNotifierItemRegistered', GLib.Variant.new('(s)', service));
+        this._nameWatcher[id] = Gio.DBus.session.watch_name(bus_name, Gio.BusNameWatcherFlags.NONE, null,
+                                                            this._itemVanished.bind(this));
+
+        this._dbusImpl.emit_property_changed('RegisteredStatusNotifierItems', GLib.Variant.new('as', this.RegisteredStatusNotifierItems));
+    },
+
+    _ensureItemRegistered: function(service, bus_name, obj_path) {
+        let id = this._getItemId(bus_name, obj_path);
+
+        if (this._items[id]) {
+            //delete the old one and add the new indicator
+            Util.Logger.warn("Attempting to re-register "+id+"; resetting instead");
+            this._items[id].reset();
+        }
+
+        this._registerItem(service, bus_name, obj_path)
+    },
+
     RegisterStatusNotifierItemAsync: function(params, invocation) {
         // it would be too easy if all application behaved the same
         // instead, ayatana patched gnome apps to send a path
         // while kde apps send a bus name
-        let service = params[0];
+        let [service] = params;
```

Comment:
Sorry, I'm not a JavaScript expert. Are these two lines equivalent?

Reply:
Yep: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

Reply:
What happens if params has two elements?